dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/route/RouteService.java

```diff
@@ -0,0 +1,162 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+import javax.annotation.Nonnull;
+import javax.servlet.http.HttpServletRequest;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.hisp.dhis.common.auth.ApiTokenAuth;
+import org.hisp.dhis.common.auth.Auth;
+import org.hisp.dhis.common.auth.HttpBasicAuth;
+import org.jasypt.encryption.pbe.PBEStringCleanablePasswordEncryptor;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.*;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StreamUtils;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class RouteService
+{
+    private final RouteStore routeStore;
+
+    private final ObjectMapper objectMapper;
+
+    @Qualifier( AES_128_STRING_ENCRYPTOR )
+    private final PBEStringCleanablePasswordEncryptor encryptor;
+
+    private static final RestTemplate restTemplate = new RestTemplate();
+
+    static
+    {
+        HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+        requestFactory.setConnectionRequestTimeout( 1_000 );
+        requestFactory.setConnectTimeout( 5_000 );
+        requestFactory.setReadTimeout( 10_000 );
+        requestFactory.setBufferRequestBody( true );
+
+        restTemplate.setRequestFactory( requestFactory );
+    }
+
+    public Route getDecryptedRoute( @Nonnull String id )
+    {
+        Route route = routeStore.getByUidNoAcl( id );
+
+        if ( route == null )
+        {
+            route = routeStore.getByUidNoAcl( id );
+        }
+
+        if ( route == null || route.isDisabled() )
+        {
+            return null;
+        }
+
+        try
+        {
+            route = objectMapper.readValue( objectMapper.writeValueAsString( route ), Route.class );
+        }
+        catch ( JsonProcessingException ex )
+        {
+            log.error( "Unable to create clone of Route with ID " + route.getUid() + ". Please check it's data." );
+            return null;
+        }
+
+        decrypt( route );
+
+        return route;
+    }
+
+    public ResponseEntity<String> exec( Route route, HttpServletRequest request )
+        throws IOException
+    {
+        HttpHeaders headers = new HttpHeaders();
+        route.getHeaders().forEach( headers::add );
+
+        if ( route.getAuth() != null )
+        {
+            route.getAuth().apply( headers );
+        }
+
+        HttpHeaders queryParameters = new HttpHeaders();
+        request.getParameterMap().forEach( ( key, value ) -> queryParameters.addAll( key, Arrays.asList( value ) ) );
```

Comment:
I think it is usually god to _not_ use `Arrays.asList` directly. It returns some sort of nested class from inside the Arrays class, and changes to this list will be reflected in the original array. Better to use a method that gives you a new, regular ArrayList, like `ListUtils.newList`.